rfcs/rfc-HAARG-0001.md

```diff
@@ -0,0 +1,180 @@
+# Map with different topic variable
+
+## Preamble
+
+    Author:  Graham Knop <haarg@haarg.org>
+    Sponsor:
+    ID: HAARG-0001
+    Status:  Draft
+
+## Abstract
+
+Allow `map` and `grep` to be given a different variable to be used as the topic
+variable, rather than using `$_`. Also allow multiple variables to be given to
+do n-at-a-time iteration.
+
+## Motivation
+
+Traditionally, `map` and `grep` loop over a list, aliasing `$_` to each list
+entry. While this can be convenient in many cases, it gets awkward when nested
+maps are needed. You need to manually save the `$_` value in another variable.
+
+```perl
+my %hash = (
+  foo => [ 1, 2, 3, 4 ],
+  bar => [ 5, 6, 7, 8 ],
+);
+
+my @out = map {
+  my $key = $_;
+  map { "$key: $_" } $hash{$key}->@*;
+} keys %hash;
+```
+
+Using `$_` can also be dangerous if you need to call code not under your
+direct control, due to it being the implicit target of operations like
+`readline`.
+
+It would be more convenient if you could use a different variable for the
+topic, similar to what `for` allows.
+
+```perl
+my @out = map my $key {
+  map my $i { "$key: $i" } $hash{$key}->@*;
+} keys %hash;
+```
+
+A natural extension of this syntax would be to allow n-at-a-time iteration, as
+`for` can do on perl 5.36+.
+
+```perl
+my @out = map my ($key, $val) {
+  map my $i { "$key: $i" } $val->@*;
+} %hash;
+```
+
+All of this would apply to `grep` as well.
+
+As the syntax proposed is currently invalid, a feature should not be needed to
+enable it.
```

Comment:
As much as I'd love to see this syntax, I have my slight doubts here. Under current 5.36:

```
eval: my @result = map my $x = $_*2, (1,2,3);  \@result
[ '2', '4', '6' ]
```

meaning that the parser is going to have to at least lookahead past the `my` and the `$x` in order to find the `{` or lack thereof. Given that you want to permit an arbitrary sized list of scalars (and maybe `undef` as well) this is a potentially long amount of lookahead. There may be further pathological cases I've not thought of.

Even if it is possible to implement it by some scary amount of code in perl's tokenizer, I'm not convinced we'd *want* to do that even so. I have a particular bee in my bonnet at the moment about trying to make Perl syntax more predictable and parsable with [simple(r) parsers] so I might be biased here... but...

I can imagine that it might require a named feature to enable this syntax while simultaneously disabling the use of `map EXPR, LIST`. I'd be happy to consider such an option. For all the new abilities such a syntax gives (as you point out, it entirely replaces `pairmap` and `pairgrep` for example), I'd be happy to have to pay for that by requesting it.

```perl
use feature 'extended_map';

my @squares = map my $x { $x **2 } @numbers;
```

I believe the RFC should at least suggest a name for a feature guard, if it turns out that one is required, so it can be discussed ahead of time rather than ad-hoc during development.

[simple(r) parsers]: https://github.com/tree-sitter-perl/tree-sitter-perl/

Reply:
My initial version of this idea did disable the EXPR form of map entirely, because I was concerned about this ambiguity. But @iabyn disagreed and said it wasn't needed. While the list of iteration variables could be unbounded, the syntax of them is simple. It doesn't require parsing perl's general and infinitely complex syntax to disambiguate, unlike the current BLOCK vs EXPR forms.

Having this feature disable the EXPR form of map would be decent compromise. But until someone proves that necessary, I don't think we need to change this in the RFC. It also means we don't need to come up with a name for the feature.

Reply:
> Having this feature disable the EXPR form of map would be decent compromise.

this would, on a long timescale, disable the expression form entirely for perl, due to social knock-on effects, and cause massive damage. it is not acceptable in any way

Reply:
also a compromise would be `:my(` syntax, as it makes the parsing wholly unambiguous and erases all aesthetic concerns about visual separation

(sorry for repost, bad github ui)